File: src/morphology.cpp
#include "conversion.h"

using namespace Rcpp;
using namespace cv;

Rcpp::NumericMatrix
imageErode(Rcpp::NumericMatrix imgMat,
	   Rcpp::NumericMatrix kern, 
	   Rcpp::IntegerVector anchor_, int iteration)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat);
    cv::Mat K = convertMat_RCPP2CV(kern);
    std::vector<int> anchor = as< std::vector<int> >(anchor_);
    cv::Mat outImg;
    cv::erode(M, outImg, K, Point(anchor[0], anchor[1]));
    return (convertMat_CV2RCPP(outImg));
}

Rcpp::NumericMatrix
imageDilate(Rcpp::NumericMatrix imgMat,
	    Rcpp::NumericMatrix kern, 
	    Rcpp::IntegerVector anchor_, int iteration)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat);
    cv::Mat K = convertMat_RCPP2CV(kern);
    std::vector<int> anchor = as< std::vector<int> >(anchor_);
    cv::Mat outImg;
    cv::dilate(M, outImg, K, Point(anchor[0], anchor[1]));
    return (convertMat_CV2RCPP(outImg));
}

RCPP_MODULE(Morph) 
{
    function("imageErode", &imageErode, "Erodes an image");
    function("imageDilate", &imageDilate, "Erodes an image");
}

File: R/morphology.R
##' .. content for \description{} (no empty lines) ..
##'
##' .. content for \details{} ..
##' @title Erodes an image using a specific structuring element
##' @param img rip object of 1 or 3 channels. 
##' @param kern structuring element
##' @param anchor position of anchor. Anchor value (-1, 1) means that the anchor is at the element center.
##' @param iter Number of times erosion to be applied
##' @return Eroded rip object of same channel and depth as img.
##' @author Kaustav Nandy
erode <- function(img, kern, anchor, iter)
{
    mmorph <- Module("Morph", PACKAGE = "RIPOpenCV")
    mmorph$imageErode(img, kern, anchor, iter)
}

##' .. content for \description{} (no empty lines) ..
##'
##' .. content for \details{} ..
##' @title Dilate an image using a specific structuring element.
##' @param img rip object of 1 or 3 channels. 
##' @param kern structuring element.
##' @param anchor position of anchor. Anchor value (-1, 1) means that the anchor is at the element center.
##' @param iter Number of times erosion to be applied
##' @return Dilated rip object of same channel and depth as img.
##' @author Kaustav Nandy
dilate <- function(img, kern, anchor, iter)
{
    mmorph <- Module("Morph", PACKAGE = "RIPOpenCV")
    mmorph$imageDilate(img, kern, anchor, iter)
}
